File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
def solution(s):
    n = len(s)
    ans = 0
    
    for i in range(n):
        # 시작점
        srt = i-1
        flg = True
        stck = []
        for j in range(n):
            srt = (srt+1) % n
            
            if( stck and
                ((stck[-1] == '(' and s[srt] == ')') or
                (stck[-1] == '[' and s[srt] == ']') or
                (stck[-1] == '{' and s[srt] == '}'))):
                stck.pop()
            elif s[srt] == '(' or s[srt] == '[' or s[srt] == '{':
                stck.append(s[srt])
            else: 
                flg = False
                break
            
        if not stck and flg == True: ans+=1
    return ans


    '''
    왼쪽으로 x칸 만큼 회전 -> s가 올바른가?
    x칸은 0~s-1 만큼 회전한것
    스택 말고 a1, a2, b1, b2, c1, c2로 하기
    
    -> 중첩구조를 파악못해서 이걸로 못품
    ([)] -> 이것도 맞는거처리되기 때문
    
    
    
    '''

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.*;
class Solution {
    public int solution(String s) {
        int answer = 0;
        
        // 왼쪽으로 s를 x칸 회전
        for(int i = 0; i < s.length(); i++){
            Stack<Integer> stck = new Stack<>();
            boolean isValid = true;
            
            // i부터 끝까지 처리
            for(int j = i; j < s.length() && isValid; j++){
                if(s.charAt(j) == '['){
                    stck.push(1);
                }else if(s.charAt(j) == '{'){
                    stck.push(2);
                }else if(s.charAt(j) == '('){
                    stck.push(3);
                }else { // 닫는 괄호들
                    if(stck.empty()) {
                        isValid = false;
                        break;
                    }
                    
                    if(s.charAt(j) == ']' && stck.peek() == 1){
                        stck.pop();
                    }else if(s.charAt(j) == '}' && stck.peek() == 2){
                        stck.pop();
                    }else if(s.charAt(j) == ')' && stck.peek() == 3){
                        stck.pop();
                    }else {
                        // 매칭되지 않는 경우
                        isValid = false;
                        break;
                    }
                }
            }
            
            // 0부터 i-1까지 처리
            for (int j = 0; j < i && isValid; j++){
                if(s.charAt(j) == '['){
                    stck.push(1);
                }else if(s.charAt(j) == '{'){
                    stck.push(2);
                }else if(s.charAt(j) == '('){
                    stck.push(3);
                }else { // 닫는 괄호들
                    if(stck.empty()) {
                        isValid = false;
                        break;
                    }
                    
                    if(s.charAt(j) == ']' && stck.peek() == 1){
                        stck.pop();
                    }else if(s.charAt(j) == '}' && stck.peek() == 2){
                        stck.pop();
                    }else if(s.charAt(j) == ')' && stck.peek() == 3){
                        stck.pop();
                    }else {
                        // 매칭되지 않는 경우
                        isValid = false;
                        break;
                    }
                }
            }
            
            if(isValid && stck.size() == 0){
                answer += 1;
            }
        }
        
        return answer;
    }
}